File: src/application/useCases/RegisterExpense.ts
import { getCurrentMonthSheetName } from '../../domain/constants/months';
import { formatBalanceMessage, parseExpenseFromText } from '../../domain/services/ExpenseParser';
import { MissingFieldsError } from '../errors';
import type { ExpenseRepository } from '../ports/ExpenseRepository';
import type { ExpenseSession } from '../ports/ExpenseSession';
import type { FileDownloader } from '../ports/FileDownloader';
import type { OcrService } from '../ports/OcrService';

export class RegisterExpense {
  constructor(
    private readonly ocr: OcrService,
    private readonly downloader: FileDownloader,
    private readonly repository: ExpenseRepository,
    private readonly session: ExpenseSession,
  ) {}

  async processImage(fileId: string, userId: string | number): Promise<void> {
    const imageBuffer = await this.downloader.download(fileId);
    const ocrText = await this.ocr.extractText(imageBuffer);
    const expense = parseExpenseFromText(ocrText);

    if (expense.missingFields.length) {
      throw new MissingFieldsError(expense.missingFields, ocrText);
    }

    this.session.set(userId, { expense, step: 'awaiting-responsible' });
  }

  async saveExpense(userId: string | number): Promise<string> {
    const current = this.session.require(userId);

    if (current.step !== 'awaiting-final') {
      throw new Error('Não encontrei uma despesa pendente. Envie o print novamente.');
    }

    const sheetName = getCurrentMonthSheetName();
    await this.repository.append(sheetName, current.expense);
    this.session.clear(userId);

    const balance = await this.repository.getMonthlyBalance(sheetName);

    if (balance !== null) {
      return `Despesa salva na aba ${sheetName}.\n\n${formatBalanceMessage(balance)}`;
    }

    return `Despesa salva na aba ${sheetName}.`;
  }
}

File: internal/application/register_expense.go
package application

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/danilomartinssilva/bot-expenses-telegram/internal/domain"
)

type RegisterExpense struct {
	Ocr        OcrService
	Downloader FileDownloader
	Repo       ExpenseRepository
	Session    ExpenseSession
}

func (r *RegisterExpense) ProcessImage(ctx context.Context, fileID string, chatID int64) error {
	image, err := r.Downloader.Download(ctx, fileID)
	if err != nil {
		return err
	}

	text, err := r.Ocr.ExtractText(ctx, image)
	if err != nil {
		return err
	}

	expense := domain.ParseExpenseFromText(text)

	if len(expense.MissingFields) > 0 {
		return &MissingFieldsError{Missing: expense.MissingFields, OcrText: text}
	}

	r.Session.Set(chatID, domain.SessionData{Expense: expense, Step: domain.StepAwaitingResponsible})
	return nil
}

func (r *RegisterExpense) SaveExpense(ctx context.Context, chatID int64) (string, error) {
	current, err := r.Session.Require(chatID)
	if err != nil {
		return "", err
	}

	if current.Step != domain.StepAwaitingFinal {
		return "", errors.New("Não encontrei uma despesa pendente. Envie o print novamente.")
	}

	sheetName := domain.CurrentMonthSheetName(time.Now())

	if err := r.Repo.Append(ctx, sheetName, current.Expense); err != nil {
		return "", err
	}

	r.Session.Clear(chatID)

	balance, ok, err := r.Repo.MonthlyBalance(ctx, sheetName)
	if err != nil {
		return "", err
	}

	message := fmt.Sprintf("Despesa salva na aba %s.", sheetName)

	if ok {
		message += "\n\n" + domain.FormatBalanceMessage(balance)
	}

	return message, nil
}
